src/operator/elementwise_binary_broadcast_op-inl.h

```diff
@@ -0,0 +1,494 @@
+        ASSIGN_DISPATCH(out, req,
+                        F<OP>(lhs.FlatTo2D<xpu, DType>(s),
+                              rhs.FlatTo2D<xpu, DType>(s)));
+      });
+    return;
+  }
+
+  TShape ret_reshaped;
+  int lhs_broadcast_axis;
+  int rhs_broadcast_axis;
+  GetBroadcastShape_(lhs.shape_, rhs.shape_, &ret_reshaped,
+                     &lhs_broadcast_axis, &rhs_broadcast_axis);
+  MSHADOW_TYPE_SWITCH(ret->type_flag_, DType, {
+      if (lhs_broadcast_axis >= 0) {
+        // broadcast lhs
+        Tensor<xpu, 1, DType> mlhs =
+            lhs.get_with_shape<xpu, 1, DType>(Shape1(lhs.shape_.Size()), s);
+        if (rhs_broadcast_axis >= 0) {
+          // broadcast both
+          Tensor<xpu, 1, DType> mrhs =
+              rhs.get_with_shape<xpu, 1, DType>(Shape1(rhs.shape_.Size()), s);
+
+          Shape<2> ret_mshape = ret_reshaped.get<2>();
+          Tensor<xpu, 2, DType> out =
+              ret->get_with_shape<xpu, 2, DType>(ret_mshape, s);
+          if (lhs_broadcast_axis == 0) {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(broadcast<0>(mlhs, ret_mshape),
+                                  broadcast<1>(mrhs, ret_mshape)));
+          } else {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(broadcast<1>(mlhs, ret_mshape),
+                                  broadcast<0>(mrhs, ret_mshape)));
+          }
+        } else {
+          // only lhs
+          Shape<3> ret_mshape = ret_reshaped.get<3>();
+          Tensor<xpu, 3, DType> out =
+              ret->get_with_shape<xpu, 3, DType>(ret_mshape, s);
+          Tensor<xpu, 3, DType> mrhs =
+              rhs.get_with_shape<xpu, 3, DType>(ret_mshape, s);
+          if (lhs.shape_.Size() == 1) {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(broadcast_scalar(mlhs, ret_mshape), mrhs));
+          } else {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(broadcast<1>(mlhs, ret_mshape), mrhs));
+          }
+        }
+      } else {
+        Tensor<xpu, 1, DType> mrhs =
+            rhs.get_with_shape<xpu, 1, DType>(mshadow::Shape1(rhs.shape_.Size()), s);
+        if (rhs_broadcast_axis >= 0) {
+          // only rhs
+          Shape<3> ret_mshape = ret_reshaped.get<3>();
+          Tensor<xpu, 3, DType> out =
+              ret->get_with_shape<xpu, 3, DType>(ret_mshape, s);
+          Tensor<xpu, 3, DType> mlhs =
+              lhs.get_with_shape<xpu, 3, DType>(ret_mshape, s);
+          if (lhs.shape_.Size() == 1) {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(mlhs, broadcast_scalar(mrhs, ret_mshape)));
+          } else {
+            ASSIGN_DISPATCH(out, req,
+                            F<OP>(mlhs, broadcast<1>(mrhs, ret_mshape)));
+          }
+        } else {
+          LOG(FATAL) << "no broadcast is needed";
+        }
+      }
+    });
+}
+
+
+template<typename xpu, typename LHS_OP, typename RHS_OP>
```

Comment:
if you feed both left hand right to both LHD_OP and RHS_OP mulbackward can be merged right? This shouldn't increase time or memory.